python/ml4ir/applications/ranking/t_test.py

```diff
@@ -88,3 +408,213 @@ def t_test_log_results(t_test_stat, pvalue, ttest_pvalue_threshold, logger):
+    Parameters
+    ----------
+    logger: Logger
+        Logger object to log t-test significance decision
+    mean: float
+        The mean of the rank differences for the entire dataset
+    variance: float
+        The variance of the rank differences for the entire dataset
+    n: int
+        The number of samples in the entire dataset
+    ttest_pvalue_threshold: float
+        P-value threshold for student t-test
+
+    Returns
+    -------
+    t_test_metrics_dict: Dictionary
+        A dictionary with the t-test metrics recorded.
+    """
+    t_test_stat, pvalue = perform_click_rank_dist_paired_t_test(mean, variance, n)
+    t_test_log_results(t_test_stat, pvalue, ttest_pvalue_threshold, logger)
+
+    # Log t-test statistic and p-value
+    t_test_metrics_dict = {'Rank distribution t-test statistic': t_test_stat,
+                           'Rank distribution t-test pvalue': pvalue,
+                           'Rank distribution t-test difference is statistically significant': pvalue < ttest_pvalue_threshold}
+    return t_test_metrics_dict
+
+
+def statistical_analysis_preprocessing(clicked_records, group_metric_running_variance_params, group_key, variance_list):
+    """
+    This function computes the mean, variance of the provided metrics for the current batch of prediction.
+    ----------
+    clicked_records: dataframe
+            A batch of the model's prediction
+
+    group_metric_running_variance_params: dict
+        A dictionary containing intermediate mean, variance and sample size for each org for each metric
+
+     group_key: list
+        The list of keys used to aggregate the metrics
+
+    variance_list: list
+            Lists of metrics for variance computation in batches
+    """
+
+    # computing batch-wise mean and variance per metric
+    metric_stat_df_list = []
+    clicked_records.set_index(group_key)
+    grouped = clicked_records.groupby(group_key)
+    for metric in variance_list:
+        grouped_agg = grouped.apply(lambda x: x[metric].mean())
+        mean_df = pd.DataFrame(
+            {str(group_key): grouped_agg.keys().values, str([metric, "mean"]): grouped_agg.values})
+
+        grouped_agg = grouped.apply(lambda x: x[metric].var())
+        var_df = pd.DataFrame({str(group_key): grouped_agg.keys().values, str([metric, "var"]): grouped_agg.values})
+
+        grouped_agg = grouped.apply(lambda x: x[metric].count())
+        count_df = pd.DataFrame(
+            {str(group_key): grouped_agg.keys().values, str([metric, "count"]): grouped_agg.values})
+
+        concat_df = pd.concat([mean_df, var_df, count_df], axis=1)
+        metric_stat_df_list.append(concat_df)
+
+    running_stats_df = pd.concat(metric_stat_df_list, axis=1)
+    running_stats_df = running_stats_df.loc[:, ~running_stats_df.columns.duplicated()]
+    # Accumulating batch-wise mean and variances
+    compute_org_running_variance_for_metrics(variance_list, group_metric_running_variance_params, running_stats_df, group_key)
```

Comment:
@mohazahran Why can't this be moved to `metrics_helper`? You just need the mean, variance and count for each group, right? Check it out. It should be just a few new line additions IMO.

Reply:
It is also doing it in batches.

Reply:
Question - after we do the stat analysis, we should I think also generate new macro means for the metrics.

As in we currently have `macro_avg_MRR` (it might be called something else, but). We would also need `macro_avg_MRR_stat_sig`